use.go

```diff
@@ -0,0 +1,464 @@
+package irutil
+
+import (
+	"fmt"
+
+	"github.com/llir/llvm/ir"
+	"github.com/llir/llvm/ir/value"
+)
+
+// User is a user of a value.
+//
+// A User has one of the following underlying types.
+//
+//    ir.Instruction   // https://godoc.org/github.com/llir/llvm/ir#Instruction
+//    ir.Terminator    // https://godoc.org/github.com/llir/llvm/ir#Terminator
+type User interface {
+	ir.LLStringer
+}
+
+// Use tracks the use of a value and its parent instruction or terminator.
+type Use struct {
+	// Use of value.
+	Val *value.Value
+	// Parent instruction or terminator which uses value as operand.
+	User User // ir.Instruction or ir.Terminator
+}
+
+// Replace replaces the use of a value with the new value.
+func (use *Use) Replace(new value.Value) {
+	*use.Val = new
+}
+
+// FuncUses returns value uses of instructions and terminators in the given
+// function.
+func FuncUses(f *ir.Func) []*Use {
+	var uses []*Use
+	for _, block := range f.Blocks {
+		for _, inst := range block.Insts {
+			us := InstUses(inst)
+			uses = append(uses, us...)
+		}
+		us := TermUses(block.Term)
+		uses = append(uses, us...)
+	}
+	return uses
+}
+
+// TODO: consider renaming InstUses to InstOperands.
+
+// InstUses returns value uses of the given instruction.
+func InstUses(inst ir.Instruction) []*Use {
```

Comment:
I have limited time right now, and may not be able to get to all the things you want me to look at.

API-wise, I wanted to note that this will result in allocation for all calls. I can imagine use cases where this might be called a lot.

The way to avoid it would be to have:

```go
func InstUses(backing []*Use, inst ir.Instruction) (out []*Use)
```

Then, append to the backing.

Then, you'll get allocation amortized over the caller. If the caller can't be bothered to set up a buffer to append to, they can just pass `nil` as the backing pointer, and `append` will allocate, as before.

The other thing in the back of my mind is the question: is there any way we can make this cheap?

Reply:
Note: `InstUses` reads to me like it will find all uses of an instruction, rather than things which the instruction makes use of. Don't have something immediately in mind which fixes that problem.

Reply:
> API-wise, I wanted to note that this will result in allocation for all calls. I can imagine use cases where this might be called a lot.

Fair point. I was quite certain this would be one of your feedbacks :) I try to strive for the most user-friendly API first, but I agree with you. Given that these functions may be called a lot, it makes sense to make it possible to avoid unnecessary allocations.

> Then, you'll get allocation amortized over the caller. If the caller can't be bothered to set up a buffer to append to, they can just pass nil as the backing pointer, and append will allocate, as before.

Sounds good. I'll update the API as suggested.

> Note: InstUses reads to me like it will find all uses of an instruction, rather than things which the instruction makes use of. Don't have something immediately in mind which fixes that problem.

I was considering to use the name `InstOperands`. The downside is that it doesn't mirror well to `FuncOperands`.

Another idea would then be to use `UsesInInst`, `UsesInFunc`. Can't say I prefer this over `InstUses`, as I think it's one of those things you get used to. However, if there is more people who prefer an alternative name to `InstUses` (and `FuncUses`), then I'd be more than happy to change it.

Also, feel free to suggest other names that I have not yet listed above. I am easily persuaded if we just have a good proposal for what to call it instead.

Reply:
> I have limited time right now, and may not be able to get to all the things you want me to look at.

No worries @pwaller, I was guessing you'd be busy during the winter holidays ^^

Just a quick question to move things forward. Are you ok with the change proposed in llir/llvm#122? If so, I'll merge that PR first, as it is a dependency of this PR.

Reply:
Thanks for the suggested change to the API @pwaller. It has now been applied.

The API has now been updated to make use of a backing slice to avoid memory allocation. To further reduce memory allocations, the use-tracking API has been updated to operate on `[]Use` instead of `[]*Use`. We still track a pointer to the used value, so this update has no effect on the usability of the API, it only helps to reduce memory allocations.